File: backend/eye-tracking/models/kalman_filter.py
"""
Kalman Filter for Gaze Position Smoothing
Reduces noise and jitter in gaze estimates
"""

import numpy as np
from filterpy.kalman import KalmanFilter
from typing import Tuple
import structlog

from src.config import settings
from lib.errors import KalmanFilterError

logger = structlog.get_logger(__name__)


class GazeKalmanFilter:
    """
    Kalman filter for 2D gaze position smoothing
    Uses constant velocity model
    """

    def __init__(self, dt: float = None):
        """
        Initialize Kalman filter

        Args:
            dt: Time step in seconds (default from settings)
        """
        self.dt = dt or settings.KALMAN_DT

        # Initialize Kalman filter
        # State: [x, y, vx, vy] (position and velocity)
        # Measurement: [x, y] (position only)
        self.kf = KalmanFilter(dim_x=4, dim_z=2)

        # State transition matrix (constant velocity model)
        self.kf.F = np.array([
            [1, 0, self.dt, 0],
            [0, 1, 0, self.dt],
            [0, 0, 1, 0],
            [0, 0, 0, 1]
        ])

        # Measurement function (observe x, y only)
        self.kf.H = np.array([
            [1, 0, 0, 0],
            [0, 1, 0, 0]
        ])

        # Process noise covariance
        q = settings.KALMAN_PROCESS_NOISE
        self.kf.Q = np.array([
            [q, 0, 0, 0],
            [0, q, 0, 0],
            [0, 0, q, 0],
            [0, 0, 0, q]
        ])

        # Measurement noise covariance
        r = settings.KALMAN_MEASUREMENT_NOISE
        self.kf.R = np.array([
            [r, 0],
            [0, r]
        ])

        # Initial state covariance
        self.kf.P *= 10

        # Initial state
        self.kf.x = np.array([0.5, 0.5, 0, 0])  # Start at screen center

        self.initialized = False
        self.measurement_count = 0

        logger.info(
            "kalman_filter_initialized",
            dt=self.dt,
            process_noise=q,
            measurement_noise=r
        )

    def update(self, measurement: Tuple[float, float], confidence: float = 1.0) -> Tuple[float, float]:
        """
        Update filter with new gaze measurement

        Args:
            measurement: (x, y) gaze coordinates (0-1 range)
            confidence: Measurement confidence (0-1), affects measurement noise

        Returns:
            Filtered (x, y) coordinates

        Raises:
            KalmanFilterError: If update fails
        """
        try:
            x, y = measurement

            # Validate input
            if not (0 <= x <= 1 and 0 <= y <= 1):
                logger.warning("invalid_gaze_measurement", x=x, y=y)
                # Clamp to valid range
                x = np.clip(x, 0, 1)
                y = np.clip(y, 0, 1)

            # Initialize filter with first measurement
            if not self.initialized:
                self.kf.x = np.array([x, y, 0, 0])
                self.initialized = True
                self.measurement_count += 1
                return x, y

            # Adjust measurement noise based on confidence
            # Lower confidence = higher noise
            r_adjusted = settings.KALMAN_MEASUREMENT_NOISE / max(confidence, 0.1)
            self.kf.R = np.array([
                [r_adjusted, 0],
                [0, r_adjusted]
            ])

            # Predict
            self.kf.predict()

            # Update with measurement
            z = np.array([x, y])
            self.kf.update(z)

            # Extract filtered position
            filtered_x = float(self.kf.x[0])
            filtered_y = float(self.kf.x[1])

            # Clamp to valid range
            filtered_x = np.clip(filtered_x, 0, 1)
            filtered_y = np.clip(filtered_y, 0, 1)

            self.measurement_count += 1

            return filtered_x, filtered_y

        except Exception as e:
            logger.error("kalman_filter_update_error", error=str(e))
            raise KalmanFilterError(f"Kalman filter update failed: {e}")

    def predict(self) -> Tuple[float, float]:
        """
        Predict next gaze position without measurement
        Useful for handling missed frames

        Returns:
            Predicted (x, y) coordinates
        """
        try:
            self.kf.predict()
            predicted_x = float(self.kf.x[0])
            predicted_y = float(self.kf.x[1])

            # Clamp to valid range
            predicted_x = np.clip(predicted_x, 0, 1)
            predicted_y = np.clip(predicted_y, 0, 1)

            return predicted_x, predicted_y

        except Exception as e:
            logger.error("kalman_filter_predict_error", error=str(e))
            raise KalmanFilterError(f"Kalman filter prediction failed: {e}")

    def get_velocity(self) -> Tuple[float, float]:
        """
        Get current estimated velocity

        Returns:
            (vx, vy) velocity in units per second
        """
        return float(self.kf.x[2]), float(self.kf.x[3])

    def get_state(self) -> np.ndarray:
        """Get full filter state"""
        return self.kf.x.copy()

    def reset(self):
        """Reset filter to initial state"""
        self.kf.x = np.array([0.5, 0.5, 0, 0])
        self.kf.P *= 10
        self.initialized = False
        self.measurement_count = 0
        logger.info("kalman_filter_reset")


class DualKalmanFilter:
    """
    Separate Kalman filters for left and right eye
    Provides independent smoothing and fusion
    """

    def __init__(self, dt: float = None):
        """Initialize dual filters"""
        self.left_filter = GazeKalmanFilter(dt)
        self.right_filter = GazeKalmanFilter(dt)
        self.fusion_weight_left = 0.5
        self.fusion_weight_right = 0.5

    def update(
        self,
        left_measurement: Tuple[float, float],
        right_measurement: Tuple[float, float],
        left_confidence: float = 1.0,
        right_confidence: float = 1.0
    ) -> Tuple[float, float]:
        """
        Update both filters and fuse results

        Args:
            left_measurement: Left eye gaze (x, y)
            right_measurement: Right eye gaze (x, y)
            left_confidence: Left eye confidence
            right_confidence: Right eye confidence

        Returns:
            Fused (x, y) coordinates
        """
        # Update individual filters
        left_filtered = self.left_filter.update(left_measurement, left_confidence)
        right_filtered = self.right_filter.update(right_measurement, right_confidence)

        # Adjust fusion weights based on confidence
        total_confidence = left_confidence + right_confidence
        if total_confidence > 0:
            self.fusion_weight_left = left_confidence / total_confidence
            self.fusion_weight_right = right_confidence / total_confidence
        else:
            self.fusion_weight_left = 0.5
            self.fusion_weight_right = 0.5

        # Fuse results
        fused_x = (
            self.fusion_weight_left * left_filtered[0] +
            self.fusion_weight_right * right_filtered[0]
        )
        fused_y = (
            self.fusion_weight_left * left_filtered[1] +
            self.fusion_weight_right * right_filtered[1]
        )

        return fused_x, fused_y

    def reset(self):
        """Reset both filters"""
        self.left_filter.reset()
        self.right_filter.reset()
        self.fusion_weight_left = 0.5
        self.fusion_weight_right = 0.5

File: electron-app/src/renderer/eye-tracking/kalman-filter.ts
/**
 * 2D Kalman Filter for gaze smoothing
 *
 * State vector: [x, y, vx, vy]
 * - (x, y): current position
 * - (vx, vy): velocity
 */

export class KalmanFilter {
  private state: number[];           // [x, y, vx, vy]
  private covariance: number[][];    // 4x4 covariance matrix
  private processNoise: number;      // Process noise (Q)
  private measurementNoise: number;  // Measurement noise (R)
  private dt: number;                // Time step (seconds)

  constructor(
    processNoise: number = 0.01,
    measurementNoise: number = 0.1,
    dt: number = 1 / 30  // 30 FPS
  ) {
    // Initialize state to zeros
    this.state = [0, 0, 0, 0];

    // Initialize covariance matrix (identity)
    this.covariance = [
      [1, 0, 0, 0],
      [0, 1, 0, 0],
      [0, 0, 1, 0],
      [0, 0, 0, 1]
    ];

    this.processNoise = processNoise;
    this.measurementNoise = measurementNoise;
    this.dt = dt;
  }

  /**
   * Predict next state based on motion model
   */
  predict(): void {
    // State transition matrix (constant velocity model)
    // x_new = x + vx * dt
    // y_new = y + vy * dt
    // vx_new = vx
    // vy_new = vy
    const F = [
      [1, 0, this.dt, 0],
      [0, 1, 0, this.dt],
      [0, 0, 1, 0],
      [0, 0, 0, 1]
    ];

    // Predict state: x = F * x
    this.state = this.matrixVectorMultiply(F, this.state);

    // Predict covariance: P = F * P * F^T + Q
    const FP = this.matrixMultiply(F, this.covariance);
    const FPFT = this.matrixMultiply(FP, this.transpose(F));

    // Add process noise
    for (let i = 0; i < 4; i++) {
      FPFT[i][i] += this.processNoise;
    }

    this.covariance = FPFT;
  }

  /**
   * Update state with measurement
   * @param measurement - [x, y] measurement
   */
  update(measurement: [number, number]): void {
    // Measurement matrix (we only measure position, not velocity)
    const H = [
      [1, 0, 0, 0],
      [0, 1, 0, 0]
    ];

    // Measurement noise covariance
    const R = [
      [this.measurementNoise, 0],
      [0, this.measurementNoise]
    ];

    // Innovation: y = z - H * x
    const Hx = this.matrixVectorMultiply(H, this.state);
    const innovation = [
      measurement[0] - Hx[0],
      measurement[1] - Hx[1]
    ];

    // Innovation covariance: S = H * P * H^T + R
    const HP = this.matrixMultiply(H, this.covariance);
    const HPHT = this.matrixMultiply(HP, this.transpose(H));
    const S = [
      [HPHT[0][0] + R[0][0], HPHT[0][1] + R[0][1]],
      [HPHT[1][0] + R[1][0], HPHT[1][1] + R[1][1]]
    ];

    // Kalman gain: K = P * H^T * S^-1
    const PHT = this.matrixMultiply(this.covariance, this.transpose(H));
    const Sinv = this.invert2x2(S);
    const K = this.matrixMultiply(PHT, Sinv);

    // Update state: x = x + K * y
    const Ky = this.matrixVectorMultiply(K, innovation);
    for (let i = 0; i < 4; i++) {
      this.state[i] += Ky[i];
    }

    // Update covariance: P = (I - K * H) * P
    const KH = this.matrixMultiply(K, H);
    const I_KH = this.subtractFromIdentity(KH);
    this.covariance = this.matrixMultiply(I_KH, this.covariance);
  }

  /**
   * Get current filtered position
   */
  getPosition(): [number, number] {
    return [this.state[0], this.state[1]];
  }

  /**
   * Get current velocity
   */
  getVelocity(): [number, number] {
    return [this.state[2], this.state[3]];
  }

  /**
   * Reset filter with new position
   */
  reset(x: number, y: number): void {
    this.state = [x, y, 0, 0];
    this.covariance = [
      [1, 0, 0, 0],
      [0, 1, 0, 0],
      [0, 0, 1, 0],
      [0, 0, 0, 1]
    ];
  }

  // ========================================================================
  // Matrix operations
  // ========================================================================

  private matrixMultiply(A: number[][], B: number[][]): number[][] {
    const rows = A.length;
    const cols = B[0].length;
    const inner = B.length;
    const result: number[][] = [];

    for (let i = 0; i < rows; i++) {
      result[i] = [];
      for (let j = 0; j < cols; j++) {
        let sum = 0;
        for (let k = 0; k < inner; k++) {
          sum += A[i][k] * B[k][j];
        }
        result[i][j] = sum;
      }
    }

    return result;
  }

  private matrixVectorMultiply(A: number[][], v: number[]): number[] {
    const result: number[] = [];
    for (let i = 0; i < A.length; i++) {
      let sum = 0;
      for (let j = 0; j < v.length; j++) {
        sum += A[i][j] * v[j];
      }
      result[i] = sum;
    }
    return result;
  }

  private transpose(A: number[][]): number[][] {
    const rows = A.length;
    const cols = A[0].length;
    const result: number[][] = [];

    for (let j = 0; j < cols; j++) {
      result[j] = [];
      for (let i = 0; i < rows; i++) {
        result[j][i] = A[i][j];
      }
    }

    return result;
  }

  private invert2x2(A: number[][]): number[][] {
    const det = A[0][0] * A[1][1] - A[0][1] * A[1][0];
    if (Math.abs(det) < 1e-10) {
      // Singular matrix, return identity
      return [[1, 0], [0, 1]];
    }

    return [
      [A[1][1] / det, -A[0][1] / det],
      [-A[1][0] / det, A[0][0] / det]
    ];
  }

  private subtractFromIdentity(A: number[][]): number[][] {
    const size = A.length;
    const result: number[][] = [];

    for (let i = 0; i < size; i++) {
      result[i] = [];
      for (let j = 0; j < size; j++) {
        if (i === j) {
          result[i][j] = 1 - A[i][j];
        } else {
          result[i][j] = -A[i][j];
        }
      }
    }

    return result;
  }
}
